### PR TITLE
[v2] Control when a Cell is executing #279

### PR DESCRIPTION
**Pull request by MarcosVn**
* Reimplements the proposed solution for the described scenario in https://github.com/datalayer/jupyter-ui/issues/279 issue (after #282 changes)

* Added the `isExecuting` attribute in `CellState` to track/control when a cell is executing;
* Also added `isAnyCellExecuting` to the cells map to control if there is any execution in progress (or if all cells are `idle`);
* Improved related comments;
* Updated `CellAdapter` to update the `isExecuting` state in new `_execute` generic function;
* Added the `CellExecuteControl.tsx` example.

### CODE REVIEW

**Comment by echarles** [comment]
Awesome!

> How do you think we could resolve this?

In this code block, https://github.com/datalayer/jupyter-ui/blob/ec9818ce698072cf79c10284102bf334edf6dbcb/packages/react/src/jupyter/kernel/KernelExecutor.ts#L216-L244 , at the end of the `error` case, I would add `this._executed.resolve(this._model);`  (no guarantee... :)

File: packages/react/src/jupyter/kernel/KernelExecutor.ts
Comment (echarles) [inline packages/react/src/jupyter/kernel/KernelExecutor.ts:265]: As discussed with @fcollonval the initial code shoudl be .

```
const { ename, evalue, traceback } = content as any as KernelMessage.IErrorMsg;
```

Could you try this way?

Comment (echarles) [reply packages/react/src/jupyter/kernel/KernelExecutor.ts:265]: @MarcosVn

Comment (MarcosVn) [reply packages/react/src/jupyter/kernel/KernelExecutor.ts:265]: @eric, my tslint configuration says that traceback does not exist in `IErrorMsg`, should I ignore this?

image

Here, content is `IReplyErrorContent` & `IExecuteCount`:

image

Comment (MarcosVn) [reply packages/react/src/jupyter/kernel/KernelExecutor.ts]: I committed it as `const { ename, evalue, traceback } = content as any as KernelMessage.IReplyErrorContent`; - it worked here.

Comment (echarles) [reply packages/react/src/jupyter/kernel/KernelExecutor.ts]: this works for me in my vscode. What about you?

```
        case 'error':
          {
            const { ename, evalue, traceback } = content as KernelMessage.IReplyErrorContent & KernelMessage.IExecuteCount;
            this._executed.reject(
              `${ename}: ${evalue}\n${(traceback ?? []).join('\n')}`
            );
          }
```

Comment (echarles) [reply packages/react/src/jupyter/kernel/KernelExecutor.ts]: message crossing. I think you can remove the intermediary `as any`

Comment (MarcosVn) [reply packages/react/src/jupyter/kernel/KernelExecutor.ts]: Updated, @echarles.

Comment (echarles) [reply packages/react/src/jupyter/kernel/KernelExecutor.ts]: :+1:  I can approve and merge unless you have more to do here. Maybe you can bump the version number to `0.16.1`  so I can release after the merge.

Comment (MarcosVn) [reply packages/react/src/jupyter/kernel/KernelExecutor.ts]: I applied the version update here @echarles . The only point is that it was on `0.17.0`, and I set it to `0.17.1`.

Thanks in advance

Comment (echarles) [reply packages/react/src/jupyter/kernel/KernelExecutor.ts]: oops I bumped at the last merge... really sorry... can you revert back to 0.17.0 ?

Comment (MarcosVn) [reply packages/react/src/jupyter/kernel/KernelExecutor.ts]: Reverted, @echarles.

**Review by echarles — APPROVED** [review]
LGTM Thx @MarcosVn

**Comment by echarles** [comment]
Merged and 0.17.0 released. Thank you so much for this @MarcosVn

**Pull request merged**
